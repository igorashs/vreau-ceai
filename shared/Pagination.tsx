import styled from 'styled-components';
import Button from '@/shared/Button';
import ArrowSvg from '@/icons/arrow.svg';
import { useEffect, useState } from 'react';
import { nanoid } from 'nanoid';

const Wrapper = styled.div`
  display: grid;
  grid-template-columns: repeat(auto-fit, 32px);
  grid-template-rows: 32px;
  align-items: center;
  justify-items: center;
  justify-content: center;
  gap: var(--baseline);

  > * {
    height: 100%;
    width: 100%;
  }

  > button:first-child {
    transform: rotate(180deg);
  }
`;

type PaginationProps = {
  min: number;
  max: number;
  onPageChange: (page: number) => void;
};

const Pagination = ({ onPageChange, min, max }: PaginationProps) => {
  const [page, setPage] = useState(min);

  useEffect(() => {
    onPageChange(page);
  }, [page]);

  return (
    <Wrapper>
      <Button
        onClick={() => setPage((n) => (n > min ? n - 1 : n))}
        icon
        disabled={page <= min}
      >
        <ArrowSvg />
      </Button>
      {[...Array(max + 1 || 0)].map((_, i) => (
        <Button
          key={nanoid()}
          btnStyle={page === i ? 'dark-text' : 'accent-text'}
          onClick={() => setPage(i)}
        >
          {i + 1}
        </Button>
      ))}
      <Button
        onClick={() => setPage((n) => (n < max ? n + 1 : n))}
        icon
        disabled={page >= max}
      >
        <ArrowSvg />
      </Button>
    </Wrapper>
  );
};

export default Pagination;
