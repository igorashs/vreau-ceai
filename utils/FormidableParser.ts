import * as validator from '@/utils/validator';
import FormidableNameSpace, { Files, File, IncomingForm } from 'formidable';
import { NextApiRequest } from 'next';
import { Formidable, ProductFields } from 'types';
import { productMessages } from './validator/schemas/product';
import { NextApiSessionRequest } from './withSession';

type ParsedData = {
  fields: Partial<ProductFields>;
  files: Files & { src?: File };
};

type HandleProductForm = (
  req: NextApiRequest | NextApiSessionRequest,
  options: Partial<FormidableNameSpace.Options>,
  allowedFileTypes: string[],
) => Promise<ParsedData>;

const handleProductForm: HandleProductForm = async (
  req,
  options,
  allowedFileTypes,
) => {
  const data: ParsedData = await new Promise((resolve, reject) => {
    const form: Formidable = new IncomingForm(options);

    form.onPart = (part) => {
      if (part.filename) {
        // get file extension
        const fileType = part.mime?.split('/').pop() ?? '';

        // validate allowed file types
        if (!allowedFileTypes.includes(fileType)) {
          form.emit(
            'error',
            validator.createValidationError({
              message: productMessages.src.invalid,
              key: 'src',
            }),
          );
        }
      }

      form.handlePart(part);
    };

    form.parse(req, (err, fields, files) => {
      if (err) {
        // custom error on maxFileSize
        if (err.message.includes('maxFileSize')) {
          reject(
            validator.createValidationError({
              message: productMessages.src.max,
              key: 'src',
            }),
          );
        } else {
          reject(err);
        }
      }

      resolve({ fields, files });
    });
  });

  return data;
};

export default {
  handleProductForm,
};
