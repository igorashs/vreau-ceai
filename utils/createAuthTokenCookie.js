import jwt from 'jsonwebtoken';
import cookie from 'cookie';

const EXPIRE_TIME = 24 * 60 * 60;

export const createAuthTokenCookie = (payload) => {
  const authToken = jwt.sign(payload, process.env.AUTH_PRIVATE_KEY, {
    expiresIn: EXPIRE_TIME
  });

  return cookie.serialize('auth-token', authToken, {
    secure: process.env.NODE_ENV !== 'development',
    httpOnly: true,
    sameSite: 'strict',
    path: '/',
    maxAge: EXPIRE_TIME
  });
};
