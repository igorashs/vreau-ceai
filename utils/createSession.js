import jwt from 'jsonwebtoken';
import cookie from 'cookie';

const EXPIRE_TIME = 24 * 60 * 60;

export const createSession = (payload) => {
  const token = jwt.sign(payload, process.env.AUTH_PRIVATE_KEY, {
    expiresIn: EXPIRE_TIME
  });

  const [header, jwtPayload, signature] = token.split('.');
  const session = header + '.' + jwtPayload;

  const sessionCookie = cookie.serialize('session', session, {
    secure: process.env.NODE_ENV !== 'development',
    sameSite: 'strict',
    path: '/',
    maxAge: EXPIRE_TIME
  });

  const authCookie = cookie.serialize('auth', signature, {
    secure: process.env.NODE_ENV !== 'development',
    httpOnly: true,
    sameSite: 'strict',
    path: '/',
    maxAge: EXPIRE_TIME
  });

  return [sessionCookie, authCookie];
};
