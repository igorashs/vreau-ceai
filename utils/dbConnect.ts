import mongoose from 'mongoose';

const dbConnect = async () => {
  if (mongoose.connection.readyState >= 1) {
    return;
  }

  await mongoose.connect(process.env.MONGODB_URI as string, {
    useNewUrlParser: true,
    useUnifiedTopology: true,
    useCreateIndex: true,
    useFindAndModify: false,
  });
};

export default dbConnect;
