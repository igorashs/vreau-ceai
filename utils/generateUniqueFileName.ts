import { nanoid } from 'nanoid';

const generateUniqueFileName = (fileName: string) => {
  // grab file extension
  const matches = fileName?.match(/\..+$/);

  // create new file name
  const uniqueFileName = nanoid() + (matches && matches[0]);

  return uniqueFileName;
};

export default generateUniqueFileName;
