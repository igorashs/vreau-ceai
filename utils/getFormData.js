export const getFormData = (data) => {
  const formData = new FormData();

  for (const name in data) {
    if (name === 'src') continue;

    formData.append(name, data[name]);
  }

  if (data.src[0]) {
    formData.append('src', data.src[0]);
  }

  return formData;
};
