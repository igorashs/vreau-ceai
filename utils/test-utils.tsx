import { render, RenderOptions } from '@testing-library/react';
import GlobalStyle from 'GlobalStyle';

const Wrapper: React.FC = ({ children }) => (
  <>
    <GlobalStyle />
    {children}
  </>
);

const customRender = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
) => render(ui, { wrapper: Wrapper, ...options });

export * from '@testing-library/react';

export { customRender as render };
