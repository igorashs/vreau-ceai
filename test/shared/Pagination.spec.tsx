import { fireEvent, render } from '@/utils/test-utils';
import Pagination from '@/shared/Pagination';

describe('Pagination', () => {
  it('renders correctly', () => {
    const min = 1;
    const max = 3;
    const currPage = 1;
    const onPageChangeMock = jest.fn();

    const { container, getByTestId } = render(
      <Pagination
        min={min}
        max={max}
        currPage={currPage}
        onPageChange={onPageChangeMock}
      />,
    );

    const firstPageBtn = getByTestId('first-page');

    expect(firstPageBtn).toHaveTextContent(min.toString());
    expect(firstPageBtn).toHaveStyleRule('color', 'var(--text-dark)');
    expect(firstPageBtn).toHaveAttribute('disabled');

    const lastPageBtn = getByTestId('last-page');

    expect(lastPageBtn).toHaveTextContent(max.toString());
    expect(lastPageBtn).toHaveStyleRule('color', 'var(--accent-text-dark)');
    expect(lastPageBtn).not.toHaveAttribute('disabled');

    expect(getByTestId('curr-page')).toHaveTextContent(currPage.toString());
    expect(getByTestId('prev-page')).toHaveAttribute('disabled');
    expect(getByTestId('next-page')).not.toHaveAttribute('disabled');

    expect(container.firstChild).toMatchSnapshot();
  });

  it("doesn't render with 1 or less max pages", () => {
    const onPageChangeMock = jest.fn();

    const { container } = render(
      <Pagination
        min={1}
        max={1}
        currPage={1}
        onPageChange={onPageChangeMock}
      />,
    );

    expect(container).toBeEmptyDOMElement();
  });

  describe('onPageChange prop', () => {
    describe('.onPrevPageClick', () => {
      it('returns prev page number', () => {
        const min = 1;
        const max = 3;
        const currPage = 2;
        const onPageChangeMock = jest.fn();

        const { getByTestId } = render(
          <Pagination
            min={min}
            max={max}
            currPage={currPage}
            onPageChange={onPageChangeMock}
          />,
        );

        fireEvent.click(getByTestId('prev-page'));

        expect(onPageChangeMock).toBeCalledTimes(1);
        expect(onPageChangeMock).toBeCalledWith(currPage - 1);
      });

      it('respects min/max props limits', () => {
        const min = 1;
        const max = 3;
        const currPage = 1;
        const onPageChangeMock = jest.fn();

        const { getByTestId, rerender } = render(
          <Pagination
            min={min}
            max={max}
            currPage={currPage}
            onPageChange={onPageChangeMock}
          />,
        );

        const prevPageBtn = getByTestId('prev-page');
        fireEvent.click(prevPageBtn);

        expect(prevPageBtn).toHaveAttribute('disabled');

        rerender(
          <Pagination
            min={min}
            max={max}
            currPage={5}
            onPageChange={onPageChangeMock}
          />,
        );

        fireEvent.click(prevPageBtn);

        expect(onPageChangeMock).toBeCalledWith(max - 1);

        rerender(
          <Pagination
            min={min}
            max={max}
            currPage={0}
            onPageChange={onPageChangeMock}
          />,
        );

        fireEvent.click(prevPageBtn);

        expect(prevPageBtn).toHaveAttribute('disabled');
        expect(onPageChangeMock).toBeCalledTimes(1);
      });
    });

    describe('.onNextPageClick', () => {
      it('returns next page number', () => {
        const min = 1;
        const max = 3;
        const currPage = 2;
        const onPageChangeMock = jest.fn();

        const { getByTestId } = render(
          <Pagination
            min={min}
            max={max}
            currPage={currPage}
            onPageChange={onPageChangeMock}
          />,
        );

        fireEvent.click(getByTestId('next-page'));

        expect(onPageChangeMock).toBeCalledTimes(1);
        expect(onPageChangeMock).toBeCalledWith(currPage + 1);
      });

      it('respects min/max props limits', () => {
        const min = 1;
        const max = 3;
        const currPage = 3;
        const onPageChangeMock = jest.fn();

        const { getByTestId, rerender } = render(
          <Pagination
            min={min}
            max={max}
            currPage={currPage}
            onPageChange={onPageChangeMock}
          />,
        );

        const nextPageBtn = getByTestId('next-page');
        fireEvent.click(nextPageBtn);

        expect(nextPageBtn).toHaveAttribute('disabled');

        rerender(
          <Pagination
            min={min}
            max={max}
            currPage={0}
            onPageChange={onPageChangeMock}
          />,
        );

        fireEvent.click(nextPageBtn);

        expect(onPageChangeMock).toBeCalledWith(min + 1);

        rerender(
          <Pagination
            min={min}
            max={max}
            currPage={5}
            onPageChange={onPageChangeMock}
          />,
        );

        fireEvent.click(nextPageBtn);

        expect(nextPageBtn).toHaveAttribute('disabled');
        expect(onPageChangeMock).toBeCalledTimes(1);
      });
    });

    describe('.onFirstPageClick', () => {
      it('returns min page number', () => {
        const min = 1;
        const max = 3;
        const currPage = 3;
        const onPageChangeMock = jest.fn();

        const { getByTestId } = render(
          <Pagination
            min={min}
            max={max}
            currPage={currPage}
            onPageChange={onPageChangeMock}
          />,
        );

        fireEvent.click(getByTestId('first-page'));

        expect(onPageChangeMock).toBeCalledTimes(1);
        expect(onPageChangeMock).toBeCalledWith(min);
      });
    });

    describe('.onLastPageClick', () => {
      it('returns max page number', () => {
        const min = 1;
        const max = 3;
        const currPage = 2;
        const onPageChangeMock = jest.fn();

        const { getByTestId } = render(
          <Pagination
            min={min}
            max={max}
            currPage={currPage}
            onPageChange={onPageChangeMock}
          />,
        );

        fireEvent.click(getByTestId('last-page'));

        expect(onPageChangeMock).toBeCalledTimes(1);
        expect(onPageChangeMock).toBeCalledWith(max);
      });
    });
  });
});
