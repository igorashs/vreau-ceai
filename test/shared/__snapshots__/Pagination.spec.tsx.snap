// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination renders correctly 1`] = `
.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  gap: 7px;
  padding: calc(var(--baseline) / 4) calc(var(--baseline) / 2);
  border-radius: 4px;
  border: 0;
  color: var(--text-light);
  background-color: var(--accent-dark);
  box-shadow: 1px 1px 2px #00000066;
  text-transform: lowercase;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  padding: 4px;
  box-shadow: none;
  padding: 0;
}

.c1:hover {
  background-color: var(--accent);
}

.c2 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  gap: 7px;
  padding: calc(var(--baseline) / 4) calc(var(--baseline) / 2);
  border-radius: 4px;
  border: 0;
  color: var(--text-light);
  background-color: var(--accent-dark);
  box-shadow: 1px 1px 2px #00000066;
  text-transform: lowercase;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  background-color: transparent;
  text-transform: initial;
  box-shadow: none;
  color: var(--text-dark);
  padding: 0;
}

.c2:hover {
  background-color: var(--accent);
}

.c2:hover {
  background-color: transparent;
  -webkit-filter: brightness(0.7);
  filter: brightness(0.7);
}

.c2:disabled {
  background-color: transparent !important;
}

.c4 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  gap: 7px;
  padding: calc(var(--baseline) / 4) calc(var(--baseline) / 2);
  border-radius: 4px;
  border: 0;
  color: var(--text-light);
  background-color: var(--accent-dark);
  box-shadow: 1px 1px 2px #00000066;
  text-transform: lowercase;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  background-color: transparent;
  text-transform: initial;
  box-shadow: none;
  color: var(--accent-text-dark);
  padding: 0;
}

.c4:hover {
  background-color: var(--accent);
}

.c4:hover {
  background-color: transparent;
  -webkit-filter: brightness(0.7);
  filter: brightness(0.7);
}

.c4:disabled {
  background-color: transparent !important;
}

.c0 {
  display: grid;
  grid-template-columns: repeat(auto-fit,24px);
  grid-template-rows: 24px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  justify-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  gap: var(--baseline);
}

.c0 > * {
  height: 24px;
  width: 24px;
}

.c0 > button:first-child {
  -webkit-transform: rotate(180deg);
  -ms-transform: rotate(180deg);
  transform: rotate(180deg);
}

.c3 {
  font-weight: 500;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  font-size: var(--h5-font-size);
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

<div
  class="c0"
>
  <button
    class="c1"
    data-testid="prev-page"
    disabled=""
  >
    <svg />
  </button>
  <button
    class="c2"
    data-testid="first-page"
    disabled=""
  >
    1
  </button>
  <p
    class="c3"
    data-testid="curr-page"
  >
    1
  </p>
  <button
    class="c4"
    data-testid="last-page"
  >
    3
  </button>
  <button
    class="c1"
    data-testid="next-page"
  >
    <svg />
  </button>
</div>
`;
